acq_stat_reports/utils.py

```diff
@@ -0,0 +1,538 @@
+    for bin_col in cols:
+        bin_edges[bin_col] = np.atleast_1d(bin_edges[bin_col])
+        if len(bin_edges[bin_col].shape) != 1:
+            raise ValueError("bin_edges must be a dict of 1D arrays")
+
+    # data = data.copy()
+    bin_edges = bin_edges.copy()
+
+    # for bin_col in cols:
+    #     # no underflow (otherwise I one has to be careful with bin == 0)
+    #     data = data[data[bin_col] > bin_edges[bin_col][0]]
+
+    sizes = [len(b) + 1 for b in bin_edges.values()]
+    offsets = {cols[0]: 1}
+    offsets.update({cols[i]: np.prod(sizes[:i]) for i in range(1, len(cols))})
+
+    global_bin_idx = np.zeros(len(data), dtype=int)
+    for bin_col in cols:
+        data[f"{bin_col}_bin"] = np.digitize(data[bin_col], bin_edges[bin_col])
+        global_bin_idx += offsets[bin_col] * data[f"{bin_col}_bin"]
+    data["bin"] = global_bin_idx
+
+    # now this is very inefficient, because we are making a list of all possible combinations of bin
+    # indices. Note that we are iterating over cols in reverse order so that the bin calculation
+    # agrees with the one on the data array.
+    bin_idx = dict(
+        zip(
+            cols[::-1],
+            np.array(
+                list(
+                    zip(
+                        *[
+                            list(j)
+                            for j in itertools.product(
+                                *[range(len(bin_edges[col]) + 1) for col in cols[::-1]]
+                            )
+                        ],
+                        strict=True,
+                    )
+                )
+            ),
+            strict=True,
+        )
+    )
+
+    # bins are the bin edges, and np.digitize returns indices assuming there are (n_bins - 1) bins,
+    # plus the under/overflow bins. That's (n_bins + 1) bins.
+    # for book-keeping, we will create padded arrays with -inf and +inf
+    padded_bins = {
+        col: np.concatenate([[-np.inf], bin_edges[col], [np.inf]]) for col in cols
+    }
+
+    quantiles = Table()
+    quantiles["bin"] = np.arange(len(bin_idx[cols[0]]))
+    for bin_col in cols:
+        j = bin_idx[bin_col]
+        quantiles[f"{bin_col}_bin"] = j
+        quantiles[f"{bin_col}_low"] = padded_bins[bin_col][j]
+        quantiles[f"{bin_col}_high"] = padded_bins[bin_col][j + 1]
+        quantiles[f"{bin_col}"] = (
+            quantiles[f"{bin_col}_low"] + quantiles[f"{bin_col}_high"]
+        ) / 2
+
+    # this is a sanity check that the bin indices as calculated above are correct
+    global_bin_idx = np.zeros(len(quantiles), dtype=int)
+    for bin_col in cols:
+        global_bin_idx += offsets[bin_col] * quantiles[f"{bin_col}_bin"]
+    assert np.all(quantiles["bin"] == global_bin_idx)
+    # end sanity check
+
+    g = data.group_by(bin_cols)
+    assert len(g.groups.indices) <= len(quantiles) + 1  # sanity check
+    idx = g[["bin"] + cols].groups.aggregate(np.mean)["bin"].astype(int)
```

Comment:
Why is this aggregating over `cols` but then ignoring the result by using only the `bin` column?

Reply:
The way the algorithm works is by assigning a unique bin ID, so when you group by `cols`, each group should have a unique bin ID, which is stored in the `bin` column.

Reply:
FWIW, this passes tests if I change the line to:
```
idx = g[["bin"]].groups.aggregate(np.mean)["bin"].astype(int)
# Or even simpler
idx = g["bin"].groups.aggregate(lambda c: c[0])
# Or even faster
idx = g["bin"][g.groups.indices[:-1]]
```